keadm/app/cmd/edge/join.go

```diff
@@ -102,14 +106,55 @@ func addJoinOtherFlags(cmd *cobra.Command, joinOptions *types.JoinOptions) {
+
+	cmd.Flags().Uint16Var(&joinOptions.WebSocketPort, types.WebSocketPort, joinOptions.WebSocketPort,
+		fmt.Sprintf("WebSocket port of kubeEdge edgecontroller, default: %v", types.DefaultWebSocketPort))
+
+	cmd.Flags().Uint16Var(&joinOptions.QuicPort, types.QuicPort, joinOptions.QuicPort,
+		fmt.Sprintf("Quic port of kubeEdge edgecontroller, default: %v", types.DefaultQuitPort))
+
+	cmd.Flags().StringVar(&joinOptions.EdgeYaml.EdgeHub.Controller.Protocal, types.Protocol, joinOptions.EdgeYaml.EdgeHub.Controller.Protocal,
+		"Protocol between edge_core and edgecontroller, support websocket and quic , default: websocket")
+
+	joinOptions.EdgeYaml.EdgeHub.WebSocket.URL = types.CreateWebSocketUrl(joinOptions.EdgeControllerIP, joinOptions.WebSocketPort, joinOptions.EdgeNodeID)
+	joinOptions.EdgeYaml.EdgeHub.Controller.NodeID = joinOptions.EdgeNodeID
+	joinOptions.EdgeYaml.EdgeD.HostnameOverride = joinOptions.EdgeNodeID
+
+	joinOptions.EdgeYaml.EdgeHub.Quic.URL = fmt.Sprintf("%s:%s", joinOptions.EdgeControllerIP, joinOptions.QuicPort)
+
+	cmd.Flags().Uint8Var(&joinOptions.EdgeYaml.MQTT.Mode, types.MqttMode, joinOptions.EdgeYaml.MQTT.Mode,
+		"Mqtt mode, support value:(0,1,2) 0: internal mqtt broker enable only. 1: internal and external mqtt broker enable. 2: external mqtt broker enable only. default: 0")
+
+	cmd.Flags().StringVar(&joinOptions.EdgeYaml.MQTT.Server, types.MqttServer, joinOptions.EdgeYaml.MQTT.Server,
+		fmt.Sprintf("set mqtt server address, such as %v , default: %v", types.DefaultMqttServer, types.DefaultMqttServer))
+
+	cmd.Flags().StringVar(&joinOptions.EdgeYaml.MQTT.InternalServer, types.MqttInternalServer, joinOptions.EdgeYaml.MQTT.InternalServer,
+		fmt.Sprintf("set mqtt internal server address, such as %v , default: %v", types.DefaultMqttInternalServer, types.DefaultMqttInternalServer))
+
 }
 
 // newJoinOptions returns a struct ready for being used for creating cmd join flags.
 func newJoinOptions() *types.JoinOptions {
-	opts := &types.JoinOptions{}
-	opts.InitOptions = types.InitOptions{DockerVersion: types.DefaultDockerVersion, KubeEdgeVersion: types.DefaultKubeEdgeVersion,
-		KubernetesVersion: types.DefaultK8SVersion}
-	opts.CertPath = types.DefaultCertPath
+	opts := &types.JoinOptions{
+		GlobalOptions: types.GlobalOptions{
+			DockerVersion:     types.DefaultDockerVersion,
+			KubeEdgeVersion:   types.DefaultKubeEdgeVersion,
+			KubernetesVersion: types.DefaultK8SVersion},
+		EdgeYaml: types.NewEdgeYamlSt(),
+		LogYaml: &types.LoggingYaml{
+			LoggerLevel:   "INFO",
+			EnableRsysLog: false,
+			LogFormatText: true,
+			Writers:       []string{"stdout"}},
+		ModulesYaml: &types.ModulesYaml{
+			Modules: types.ModulesSt{
+				Enabled: []string{"eventbus", "servicebus", "websocket", "metaManager", "edged", "twin"}},
+		},
+		//If the user doesn't provide any edge ID on the command line, then it generates unique id and assigns it.
+		EdgeNodeID:    uuid.New().String(),
+		WebSocketPort: 10000,
+		QuicPort:      10001,
```

Comment:
Replace these with the newly defined constants @ keadm/app/cmd/common/constant.go